conf/gulp-tasks/bundle/bundleabstractfactory.js

```diff
@@ -0,0 +1,188 @@
+const replace = require('gulp-replace');
+
+// An enum describing the different types of SDK bundle.
+const BundleType = {
+  MODERN: 'modern',
+  LEGACY_IIFE: 'legacy-iife',
+  LEGACY_UMD: 'legacy-umd'
+};
+Object.freeze(BundleType);
+
+/**
+ * A class providing factories for building different types of SDK bundle.
+ */
+class BundleAbstractFactory {
+  constructor () {
+    this._namespace = 'ANSWERS';
+  }
+
+  /**
+   * Creates a factory for building a specific type of SDK bundle.
+   *
+   * @param {BundleType} bundleType The type of SDK bundle to build.
+   * @returns {Function} Factory method for building an SDK bundle of the
+   *                     the specified type.
+   */
+  create (bundleType) {
+    switch (bundleType) {
+      case BundleType.MODERN:
+        return () => this._modernBundle();
+      case BundleType.LEGACY_IIFE:
+        return () => this._legacyBundleIIFE();
+      case BundleType.LEGACY_UMD:
+        return () => this._legacyBundleUMD();
+      default:
+        throw new Error('Unrecognized BundleType');
+    }
+  }
+
+  /**
+   * A factory method for producing the modern SDK bundle.
+   *
+   * @returns {File} The modern SDK bundle.
+   */
+  _modernBundle () {
+    const rollupConfig = {
+      input: './src/answers-umd.js',
+      output: {
+        format: 'umd',
+        name: this._namespace,
+        exports: 'default',
+        sourcemap: true
+      },
+      plugins: [
+        resolve(),
+        commonjs({
+          include: './node_modules/**'
+        }),
+        babel({
+          babelrc: false,
+          exclude: 'node_modules/**',
+          presets: ['@babel/env']
+        })
+      ]
+    };
+    return rollup(rollupConfig)
+      .pipe(source('answers-modern.js'))
+      .pipe(replace('@@LIB_VERSION', this._getLibVersion()))
+      .pipe(dest('dist'));
+  }
+
+  /**
+   * A factory method for producing the legacy, IIFE-style SDK bundle.
+   *
+   * @returns {File} The legacy, IIFE-style SDK bundle.
+   */
+  _legacyBundleIIFE () {
+    const legacyBundleConfig = {
+      format: 'iife',
+      name: this._namespace,
+      sourcemap: true
+    };
+    return this._legacyBundle(legacyBundleConfig, 'answers.js');
+  }
+
+  /**
+   * A factory method for producing the legacy, UMD-style SDK bundle.
+   *
+   * @returns {File} The legacy, UMD-style SDK bundle.
+   */
+  _legacyBundleUMD () {
+    const legacyBundleConfig = {
+      format: 'umd',
+      name: this._namespace,
+      export: 'default',
+      sourcemap: true
+    };
+    return this._legacyBundle(legacyBundleConfig, 'answers-umd.js');
+  }
+
+  /**
+   * A helper method that can produce either variant of the legacy SDK bundle.
+   *
+   * @param {Object<string, ?>} outputConfig Any variant-specific configuration
+   *                                         for the legacy bundle.
+   * @param {string} fileName The filename of the created bundle.
+   * @returns {File} The legacy SDK bundle.
```

Comment:
I don't think this returns a file, it looks like it returns whatever .pipe() returns, which it looks like is a stream.Writable? (not too familiar with node streams)